From 5730c0baf57b7e1f7b0ef3ac6b16c3ed40ee8cb0 Mon Sep 17 00:00:00 2001
From: c4pp4
Date: Thu, 9 Jan 2020 02:53:47 +0100
Subject: [PATCH 1/1] Use unity-control-center as the fallback

Try to use unity-control-center as the fallback when gnome-control-center
is not available.

Signed-off-by: c4pp4
---
 src/utils/gcal-utils.c | 49 +++++++++++++++++++++++++++++++++++++++++-
 1 file changed, 48 insertions(+), 1 deletion(-)

diff --git a/src/utils/gcal-utils.c b/src/utils/gcal-utils.c
index 995449d..e17c360 100644
--- a/src/utils/gcal-utils.c
+++ b/src/utils/gcal-utils.c
@@ -1202,6 +1202,49 @@ gcal_translate_child_window_position (GtkWidget *target,
   return TRUE;
 }
 
+static gboolean
+in_desktop (const gchar *name)
+{
+  const gchar *desktop_name_list;
+  gchar **names;
+  gboolean in_list = FALSE;
+  gint i;
+
+  desktop_name_list = g_getenv ("XDG_CURRENT_DESKTOP");
+  if (!desktop_name_list)
+    return FALSE;
+
+  names = g_strsplit (desktop_name_list, ":", -1);
+  for (i = 0; names[i] && !in_list; i++)
+    if (strcmp (names[i], name) == 0) {
+      in_list = TRUE;
+      break;
+    }
+  g_strfreev (names);
+
+  return in_list;
+}
+
+static void
+spawn_uoa_with_args (const gchar *action,
+                     const gchar *arg)
+{
+  const gchar* const command[] = { "unity-control-center",
+                                   "online-accounts",
+                                   action, arg, NULL };
+  GError *error = NULL;
+
+  g_spawn_async (NULL, (gchar**) command,
+                 NULL,
+                 G_SPAWN_SEARCH_PATH | G_SPAWN_STDOUT_TO_DEV_NULL,
+                 NULL, NULL, NULL, &error);
+
+  if (error != NULL) {
+    g_warning ("Couldn't open Online Accounts panel: %s", error->message);
+    g_error_free (error);
+  }
+}
+
 void
 gcal_utils_launch_online_accounts_panel (GDBusConnection *connection,
                                          const gchar     *action,
@@ -1244,7 +1287,11 @@ gcal_utils_launch_online_accounts_panel (GDBusConnection *connection,
 
   if (!proxy)
     {
-      g_warning ("Couldn't open Online Accounts panel");
+      if (in_desktop("Unity"))
+        spawn_uoa_with_args (action, arg);
+      else
+        g_warning ("Couldn't open Online Accounts panel");
+
       return;
     }
 
-- 
2.24.1

